annotations/service.go

```diff
@@ -0,0 +1,529 @@
+		tx.Rollback()
+		return nil, err
+	}
+
+	// upsert each stream individually. a possible enhancement might be to do this as a single batched query
+	// it is unlikely that this would offer much benefit since there is currently no mechanism for creating large numbers
+	// of annotations simultaneously
+	now := time.Now()
+	for name := range streams {
+		query, args, err := newUpsertStreamQuery(orgID, s.idGenerator.ID(), now, influxdb.Stream{Name: name})
+		if err != nil {
+			tx.Rollback()
+			return nil, err
+		}
+
+		var streamID platform.ID
+		if err = tx.GetContext(ctx, &streamID, query, args...); err != nil {
+			tx.Rollback()
+			return nil, err
+		}
+
+		streams[name] = streamID
+	}
+
+	// bulk insert for the creates. this also is unlikely to offer much performance benefit, but since the query
+	// is only used here it is easy enough to form to bulk query.
+	q := sq.Insert("annotations").
+		Columns("id", "org_id", "stream_id", "name", "summary", "message", "stickers", "duration", "lower", "upper").
+		Suffix("RETURNING *")
+
+	for _, create := range creates {
+		// double check that we have a valid name for this stream tag - error if we don't. this should never be an error.
+		streamID, ok := streams[create.StreamTag]
+		if !ok {
+			tx.Rollback()
+			return nil, &ierrors.Error{
+				Code: ierrors.EInternal,
+				Msg:  fmt.Sprintf("unable to find id for stream %q", create.StreamTag),
+			}
+		}
+
+		// add the row to the query
+		newID := s.idGenerator.ID()
+		lower := create.StartTime.Format(time.RFC3339Nano)
+		upper := create.EndTime.Format(time.RFC3339Nano)
+		duration := timesToDuration(*create.StartTime, *create.EndTime)
+		q = q.Values(newID, orgID, streamID, create.StreamTag, create.Summary, create.Message, create.Stickers, duration, lower, upper)
+	}
+
+	// get the query string and args list for the bulk insert
+	query, args, err := q.ToSql()
+	if err != nil {
+		tx.Rollback()
+		return nil, err
+	}
+
+	// run the bulk insert and store the result
+	var res []influxdb.StoredAnnotation
+	if err := tx.SelectContext(ctx, &res, query, args...); err != nil {
+		tx.Rollback()
+		return nil, err
+	}
+
+	if err = tx.Commit(); err != nil {
+		return nil, err
+	}
+
+	// convert the StoredAnnotation structs to AnnotationEvent structs before returning
+	return storedAnnotationsToEvents(res)
+}
+
+// ListAnnotations returns a list of annotations from the database matching the filter
+// For time range matching, sqlite is able to compare times with millisecond accuracy
+func (s *Service) ListAnnotations(ctx context.Context, orgID platform.ID, filter influxdb.AnnotationListFilter) ([]influxdb.StoredAnnotation, error) {
+	// we need to explicitly format time strings here and elsewhere to ensure they are `
+	// by the database consistently
+	sf := filter.StartTime.Format(time.RFC3339Nano)
+	ef := filter.EndTime.Format(time.RFC3339Nano)
+
+	q := sq.Select("annotations.*").
+		Distinct().
+		From("annotations, json_each(stickers) AS json").
```

Comment:
My SQL is rusty... is there an implicit join here? I'm not seeing how the filter on `json.value` propagates to the rows in `annotations`

Reply:
Yeah it seems like you could think of it as an implicit join. The [docs](https://www.sqlite.org/json1.html) describe it as a table-value function. I thought [this blog post](http://blog.benjamin-encz.de/post/sqlite-one-to-many-json1-extension/) was helpful to understand how it worked.

Reply:
Aha, I missed that `stickers` is a column and not a table.

Does the query also work if you fully specify `annotations.stickers` here? I think it'd be more readable at a glance.

Reply:
Yep that seems to work! I made the change so it says `annotations.stickers` to help with readability.